dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/OpenTelemetryInfoFactory.java

```diff
@@ -0,0 +1,358 @@
+
+    private static final Map<J2EEName, OpenTelemetryInfo> appToInfo = new ConcurrentHashMap<J2EEName, OpenTelemetryInfo>();
+    private static final Set<J2EEName> shutDownApplications = Collections.newSetFromMap(new WeakHashMap<J2EEName, Boolean>()); //We have no good place to say forget about an app entirely, so I use weakreferences. 
+
+    //The key methods that provide access to telemetry objects begin here
+    public static OpenTelemetryInfo getOpenTelemetryInfo(J2EEName j2EEName) {
+        return appToInfo.computeIfAbsent(j2EEName, OpenTelemetryInfoFactory::createOpenTelemetryInfo);
+    }
+
+    public static Tracer getTracer(J2EEName j2EEName) {
+        try {
+            return getOpenTelemetryInfo(j2EEName).getOpenTelemetry().getTracer(INSTRUMENTATION_NAME);
+        } catch (Exception e) {
+            Tr.error(tc, Tr.formatMessage(tc, "CWMOT5002.telemetry.error", e));
+            return OpenTelemetry.noop().getTracerProvider().get("");
+        }
+    }
+
+    public static Span getSpan() {
+        return new SpanProxy(); //TODO I cannot see a way to get no-op spans or baggage, check if we need code here to handle apps that have shut down.
+    }
+
+    public static Baggage getBaggage() {
+        return new BaggageProxy();
+    };
+
+    //The key methods that provide access to telemetry objects end here
+
+    private static OpenTelemetryInfo createOpenTelemetryInfo(J2EEName j2EEName) {
+        //If the app has already shut down but something still tries to get OpenTelemetry objects return a no-op object
+        if (shutDownApplications.contains(j2EEName)){ //There is a low risk race condition here. Thread A gets past this check just as thread B shuts everything down.
+            Tr.warning(tc, Tr.formatMessage(tc, "CWMOT5003.factory.used.after.shutdown"));
+            return new OpenTelemetryInfo(false, OpenTelemetry.noop());
+        }
+
+        try {
+            if (AgentDetection.isAgentActive()) {
+                // If we're using the agent, it will have set GlobalOpenTelemetry and we must use its instance
+                // all config is handled by the agent in this case
+                return new OpenTelemetryInfo(true, GlobalOpenTelemetry.get());
+            }
+
+            final Map<String, String> telemetryProperties = getTelemetryProperties();
+
+            //Builds tracer provider if user has enabled tracing aspects with config properties
+            if (!checkDisabled(telemetryProperties)) {
+                OpenTelemetry openTelemetry = AccessController.doPrivileged( (PrivilegedAction<OpenTelemetry>) () -> {
+                    //This contains API calls that change between the upstream open telemetry version.
+                    //We use @Reference because the callers coming into  OpenTelemetryInfoFactory via
+                    //The static accessor do not know which version of mpTelemetry will be in use.
```

Comment:
Out of date comment here, we no longer use `@Reference`

Reply:
Done